### PR TITLE
Enhance date parser

### LINKED ISSUE #3456: Support single number datetime/date/time parsing
**Introduction**

In function calling as bellow:
```
date('2019')
datetime('2020')
```
In current, we only parse the string '2019' and '2020', so parser can't determine the deduce rule and lead to confliction for losing the information. So we could insert the function call information back to the raw string to fix it.
e.g. date('2019') we parse '2019' now, could change to parse '__date2019'.

**Contents**

**Related work**

### PR DESCRIPTION
**Pull request by jackwener**
#### What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### What problem(s) does this PR solve?
Issue(s) number:  close #3456

Description:

#### Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/common/time/parser/datetime_parser.yy [hunk 1]
```diff
@@ -78,34 +78,43 @@ enum class Type {
 %%
 
 datetime
-  : date date_time_delimiter time opt_time_zone {
+  : DATETIMEPRFEIX date date_time_delimiter time opt_time_zone {
```
Comment (Shylock-Hg) [inline src/common/time/parser/datetime_parser.yy:81]: Now `date_time_delimiter` and `time` could be optional.

File: src/common/time/parser/datetime_parser.yy [hunk 2]
```diff
@@ -78,34 +78,43 @@ enum class Type {
...
     if (outputType != nebula::time::Type::kDateTime) {
```
Comment (Shylock-Hg) [inline src/common/time/parser/datetime_parser.yy:82]: This could be removed now.

File: src/common/time/parser/datetime_parser.yy [hunk 3]
```diff
@@ -78,34 +78,43 @@ enum class Type {
...
-    $$ = new DateTime(TimeConversion::dateTimeShift(DateTime(*$1, *$3), -$4));
+    $$ = new DateTime(TimeConversion::dateTimeShift(DateTime(*$2, *$4), -$5));
     *output = $$;
-    delete $1;
-    delete $3;
+    delete $2;
+    delete $4;
+  }
+  | DATETIMEPRFEIX date {
+    if (outputType != nebula::time::Type::kDateTime) {
```
Comment (Shylock-Hg) [inline src/common/time/parser/datetime_parser.yy:93]: ditto.

File: src/common/time/parser/datetime_parser.yy [hunk 4]
```diff
@@ -78,34 +78,43 @@ enum class Type {
...
+      delete $2;
+      throw DatetimeParser::syntax_error(@2, "Mismatched date time type.");
+    }
+    $$ = new DateTime(*$2);
+    *output = $$;
+    delete $2;
   }
-  | date {
+  | DATEPRFEIX date {
     if (outputType != nebula::time::Type::kDate) {
```
Comment (Shylock-Hg) [inline src/common/time/parser/datetime_parser.yy:102]: ditto.

File: src/common/time/parser/datetime_parser.yy [hunk 5]
```diff
@@ -78,34 +78,43 @@ enum class Type {
...
     }
-    $$ = new DateTime(*$1);
+    $$ = new DateTime(*$2);
     *output = $$;
-    delete $1;
+    delete $2;
   }
-  | time opt_time_zone {
+  | TIMEPRFEIX time opt_time_zone {
     if (outputType != nebula::time::Type::kTime) {
```
Comment (Shylock-Hg) [inline src/common/time/parser/datetime_parser.yy:111]: ditto.

File: src/common/time/parser/datetime_parser.yy [hunk 6]
```diff
@@ -59,7 +59,7 @@ enum class Type {
 %token KW_TIME_ID
 
 /* symbols */
-%token TIME_DELIMITER SPACE POSITIVE NEGATIVE
+%token TIME_DELIMITER SPACE POSITIVE NEGATIVE DATETIMEPRFEIX DATEPRFEIX TIMEPRFEIX
```
Comment (Shylock-Hg) [inline src/common/time/parser/datetime_parser.yy:62]: KW_DATETIME is better.

**Review by Shylock-Hg — APPROVED** [review]

**Review by CPWstatic — APPROVED** [review]

**Pull request merged**